src/json.rs

```diff
@@ -98,13 +97,115 @@ fn can_get_selected_value() {
     assert_eq!(result, Some(&Value::Bool(true)));
 }
 
-/// Create [`TreeItem`]s from a [JSON](Value).
-#[must_use]
-pub fn tree_items(root: &Value) -> Vec<TreeItem<'_, Selector>> {
-    match root {
-        Value::Object(object) => from_object(object),
-        Value::Array(array) => from_array(array),
-        _ => vec![TreeItem::new_leaf(Selector::None, get_value_span(root))],
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug, Clone)]
+pub struct JsonTreeItem<'json> {
+    key: Selector,
+    value: &'json Value,
+    children: Vec<Self>,
+}
+
+impl<'json> JsonTreeItem<'json> {
+    const fn raw_leaf(key: Selector, value: &'json Value) -> Self {
+        Self {
+            key,
+            value,
+            children: Vec::new(),
+        }
+    }
+
+    const fn raw(identifier: Selector, value: &'json Value, children: Vec<Self>) -> Self {
+        // no need for child identifier uniqueness check as JSON itself already prevents this
+        Self {
+            key: identifier,
+            value,
+            children,
+        }
+    }
+
+    #[must_use]
+    pub fn new(root: &'json Value) -> Vec<Self> {
+        match root {
+            Value::Array(array) => Self::from_array(array),
+            Value::Object(object) => Self::from_object(object),
+            _ => vec![Self::raw_leaf(Selector::None, root)],
+        }
+    }
+
+    #[must_use]
+    fn new_recurse(key: Selector, value: &'json Value) -> Self {
+        match value {
+            Value::Object(object) if object.is_empty() => Self::raw_leaf(key, value),
+            Value::Object(object) => Self::raw(key, value, Self::from_object(object)),
+            Value::Array(array) if array.is_empty() => Self::raw_leaf(key, value),
+            Value::Array(array) => Self::raw(key, value, Self::from_array(array)),
+            _ => Self::raw_leaf(key, value),
+        }
+    }
+
+    #[must_use]
+    fn from_object(object: &'json serde_json::Map<String, Value>) -> Vec<Self> {
```

Comment:
Before this wasn't part of any `impl` so it just moved. I don't think people would need to use this so its an internal helper, and it's probably easier to keep generics and traits out of it to get the simplest result? Implementing exactly what's needed rather than trying to get generics and traits in should also be easier to optimize.